core/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java

```diff
@@ -0,0 +1,399 @@
+import org.codehaus.mojo.animal_sniffer.IgnoreJRERequirement;
+
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/8024")
+/**
+ * For Android users: this class is only supported in API level 24 and above.
+ * AdvancedTlsX509TrustManager is an {@code X509ExtendedTrustManager} that allows users to configure
+ * advanced TLS features, such as root certificate reloading, peer cert custom verification, etc.
+ */
+@IgnoreJRERequirement
+public final class AdvancedTlsX509TrustManager extends X509ExtendedTrustManager {
+  private static final Logger log = Logger.getLogger(AdvancedTlsX509TrustManager.class.getName());
+
+  private Verification verification;
+  private PeerVerifier peerVerifier;
+  private SslSocketPeerVerifier sslSocketPeerVerifier;
+  private SslEnginePeerVerifier sslEnginePeerVerifier;
+  // The trust certs we will use to verify peer certificate chain.
+  private volatile X509Certificate[] trustCerts;
+
+  private AdvancedTlsX509TrustManager(Verification verification,  PeerVerifier peerVerifier,
+      SslSocketPeerVerifier socketPeerVerifier, SslEnginePeerVerifier enginePeerVerifier) {
+    this.verification = verification;
+    this.peerVerifier = peerVerifier;
+    this.sslSocketPeerVerifier = socketPeerVerifier;
+    this.sslEnginePeerVerifier = enginePeerVerifier;
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, null, false);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, socket, false);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+      throws CertificateException {
+    checkTrusted(chain, authType, engine, null, false);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain,  String authType, SSLEngine engine)
+      throws CertificateException {
+    checkTrusted(chain, authType, engine, null, true);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, null, true);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, socket, true);
+  }
+
+  @Override
+  public X509Certificate[] getAcceptedIssuers() {
+    return trustCerts;
+  }
+
+  // If this is set, we will use the default trust certificates stored on user's local system.
+  // After this is used, functions that will update this.trustCerts(e.g. updateTrustCredentials(),
+  // updateTrustCredentialsFromFile()) should not be called.
+  public void useSystemDefaultTrustCerts() {
+    this.trustCerts = new X509Certificate[0];
```

Comment:
How does "empty array" become "system default"? I didn't think you use an empty keystore to initialize system defaults.

Reply:
Oh yes, empty array doesn't make it to use the system default, but I think calling something like `trustManagerFactory.init(null)` does? I've made the corresponding changes.
My reference are mainly this [SO answer](https://stackoverflow.com/questions/31292322/which-keystore-is-used-when-null-is-passed-to-keymanagerfactory-init-in-java#:~:text=1%20Answer&text=When%20null%20is%20used%20as,using%20an%20empty%20list%20internally).) and [this section of the official doc](https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html#DefaultKTMCust).